src/core/BigDecimal.ts

```diff
@@ -0,0 +1,326 @@
+     * digits of this BigDecimal.
+     */
+    readonly unscaledValue: BigInt;
+
+    /**
+     * Scale of this `BigDecimal`.  If zero or positive, the scale
+     * is the number of digits to the right of the decimal point.
+     * If negative, the unscaled value of the number is multiplied
+     * by ten to the power of the negation of the scale. For example,
+     * a scale of `-3` means the unscaled value is multiplied by 1000.
+     */
+    readonly scale: number;
+
+    /** @internal */
+    private constructor(unscaledValue: BigInt, scale: number) {
+        this.unscaledValue = unscaledValue;
+        this.scale = scale;
+    }
+
+    /** @internal */
+    private static parseExp(input: string, offset: number, len: number): number {
+        let exp = 0;
+        offset++;
+        let c = input[offset];
+        len--;
+        const negexp = (c === '-');
+        if (negexp || c === '+') {
+            offset++;
+            c = input[offset];
+            len--;
+        }
+        if (len <= 0) {
+            throw new RangeError('No exponent digits');
+        }
+        while (len > BigDecimal.MAX_EXPONENT_DIGITS && c === '0') {
+            offset++;
+            c = input[offset];
+            len--;
+        }
+        if (len > BigDecimal.MAX_EXPONENT_DIGITS) {
+            throw new RangeError('Too many nonzero exponent digits');
+        }
+        for (; ; len--) {
+            let v: number;
+            if (c >= '0' && c <= '9') {
+                v = +c;
+            } else {
+                throw new RangeError('Not a digit.');
+            }
+            exp = exp * 10 + v;
+            if (len === 1) {
+                break;
+            }
+            offset++;
+            c = input[offset];
+        }
+        if (negexp) {
+            exp = -exp;
+        }
+        return exp;
+    }
+
+    /** @internal */
+    private static adjustScale(scl: number, exp: number): number {
+        const adjustedScale = scl - exp;
+        if (!Number.isSafeInteger(adjustedScale)) {
+            throw new RangeError('Scale out of range.');
+        }
+        scl = adjustedScale;
+        return scl;
+    }
+
+    /** @internal */
+    static _new(bigDecimalString: string): BigDecimal {
+        if (typeof bigDecimalString !== 'string') {
+            throw new RangeError('String value expected');
+        }
+        let offset = 0;
+        let len = bigDecimalString.length;
+        let precision = 0;
+        let scale = 0;
+        let bigIntValue: BigInt | null = null;
+
+        let isneg = false;
+        if (bigDecimalString[offset] === '-') {
+            isneg = true;
+            offset++;
+            len--;
+        } else if (bigDecimalString[offset] === '+') {
+            offset++;
+            len--;
+        }
+
+        let dot = false;
+        let exp = 0;
+        let c: string;
+        let idx = 0;
+        const coeff = [];
+        for (; len > 0; offset++, len--) {
```

Comment:
I think while loop might be more readable but I am fine with this

Reply:
I think this is fine since for loop is too long